Download link function

rmbielby opened the pull request:
## Overview of changes

Adding a standard way to show downloadable content in the form of `download_link()`. I've based it heavily on a combination of `shiny::downloadLink()` and `shinyGovstyle::external_link()`. As with `shiny::downloadLink()`, it works with `shiny::downloadHandler()`, but had the added options to explicitly provide the file type and file size.

This is the resulting html when using `download_link()`:

`<a id="download_data" class="shiny-download-link shiny-bound-output" href="session/1492a1e83012ff241b242c018e6b28b6/download/download_data?w=" target="_blank" download="" aria-live="polite">Download data (CSV, 1 KB)</a>`

Which looks like this in the app (note the **download_link** is just the heading in the demo app):

<img width="189" alt="image" src="https://github.com/user-attachments/assets/d208fa90-5a36-43c2-8267-f44cef3d6fa2" />

Note that `aria-live="polite"` is still set, but I think this is the right way to go. For example if the user set's the link_text or file_size to update automatically based on some dynamic selections, then we'd want that to be announced at the "the next graceful opportunity".

I've not included a dynamic way to do the file sizes in this, but user's can just do their own workaround by using a reactive that evaluates a predicted file size (e.g. `table_size_in_mb <- reactive(...)`) and then calling the download link as `download_link("table_data", "Download table data", file_size = table_size_in_mb())`. Admittedly the `...` in the above is doing a lot of heavy lifting in that description, but to be fair, shinyGovstyle is about styling rather than writing active code for user's, so maybe fulfilling that part would be overreach anyway.

Fixes some of #127 

## PR Checklist

- [x] I have updated the documentation (if needed)
- [x] I have added or updated tests for these changes (if applicable)
- [x] I have tested these changes locally using `devtools::check()`

## Reviewer instructions

<!-- Put any specific questions, instructions or notes for reviewers in here -->

cjrace commented on R/download_link.R:
```diff
@@ -0,0 +1,130 @@
+#' @param outputId The name of the output slot that the `downloadHandler` is assigned
+#' to.
+#' @param link_text Text that will appear describing the download action
+#' (default: "Download file").
```
Default looks to be "Download data" not "Download file"?

rmbielby replied:
I've removed the default now anyway, based on a later comments

cjrace commented on R/download_link.R:
```diff
@@ -0,0 +1,130 @@
+#'
+#' @examples
+#' \dontrun{
+#' ui <- fluidPage(
+#'   p("Choose a dataset to download."),
+#'   selectInput("dataset", "Dataset", choices = c("mtcars", "airquality")),
+#'   download_link("download_data", "Download data")
+#' )
+#'
+#' server <- function(input, output) {
+#'   # The requested dataset
+#'   data <- reactive({
+#'     get(input$dataset)
+#'   })
+#'
+#'   output$download_data <- downloadHandler(
+#'     filename = function() {
+#'       # Use the selected dataset as the suggested file name
+#'       paste0(input$dataset, ".csv")
+#'     },
+#'     content = function(file) {
+#'       # Write the dataset to the `file` that will be downloaded
+#'       write.csv(data(), file)
+#'     }
+#'   )
+#' }
+#'
+#' shinyApp(ui, server)
```
To be consistent with elsewhere I'd wrap in `if(interactive()) {}` instead of `\dontrun{}` (e.g. header.R)

rmbielby replied:
Done

cjrace replied:
So good you did it twice...

cjrace commented on R/download_link.R:
```diff
@@ -0,0 +1,130 @@
+#' @returns shiny.tag object
+#' @export
+#'
+#' @examples
```
I think it would be worth including an example showing the file size (either in the example you have, or as an additional one)

cjrace commented on R/download_link.R:
```diff
@@ -0,0 +1,130 @@
+#' Download link
+#'
```
Could we add a note explaining the importance of including the file size?

cjrace commented on R/download_link.R:
```diff
@@ -0,0 +1,130 @@
+#' }
+download_link <- function(
+    outputId,
+    link_text = "Download data",
```
Does this not fail our own bad links checks? (if not, should it? It doesn't feel particularly descriptive)

rmbielby replied:
It's a fair cop. I've added `download data` to the banned list now and removed any default as well. If we're wanting to avoid generic text, I guess it's better not to give users an easy way out with a default value.

cjrace commented on tests/testthat/test-download_link.R:
```diff
@@ -0,0 +1,61 @@
+  expect_error(download_link("download_data", "www.google.com"))
+})
+
+test_that("Surrounding whitespace shrubbery is trimmed", {
```
![image](https://github.com/user-attachments/assets/9a14858b-189c-4c41-a946-bce9907b7a1d)

cjrace commented on R/download_link.R:
```diff
@@ -0,0 +1,130 @@
+  # file, such as the CSV version of a table in an app.
+  if (!is.null(file_size)) {
+    if(!grepl("KB$|MB$|GB$| rows$", file_size)){
+      error("File size should be a string ending in one of KB, MB, GB or rows.")
```
Is the error function from a package? Could we explicitly state it 
![image](https://github.com/user-attachments/assets/2a957413-2941-448e-9ef5-ed2145f49065)

Or did you mean `stop()`? Actually, I think you did probably mean `stop()` as that's how we do this elsewhere?

cjrace commented on inst/example_app/app.R:
Worth adding a section for downloads into the README? Currently we have examples of most of our functions in there as well as the example app (though we could also decide to cut back from the README and lean on the app more if it feels too duplicative)

rmbielby replied:
Added a small section. I would definitely vote in favour of deleting pretty much all of the example function reference from the readme though and just have people use the auto-generated reference. Feels like we're double maintaining.

cjrace replied:
Raised #129 to cover this separately

cjrace reviewed (changes requested):
Nice! Honestly quite a neat solution 😄 

To be fair, I've just looked around all over to try to find a reference point for this, and I can't find one anywhere either. So given that, I think I agree with your comment about shinyGovstyle mainly being for styling, so with a few tweaks to the function documentation I'd say this PR could fully resolve #127.

One other thought I'd had when looking through is that there is another type of download pattern that we could account for - the radio button file type selection with a confirmation button to download, like how EES does it in the table tool? ([example featured table, you'll see if you scroll to the bottom](https://explore-education-statistics.service.gov.uk/data-tables/fast-track/4bf00b2f-dec4-43da-ec12-08dcdd324466))
![image](https://github.com/user-attachments/assets/831c09d2-c660-49d9-a090-294c45a833b2)

I have done a version of this in the [provider dashboard in DfE](https://github.com/dfe-analytical-services/apprenticeships-provider-dashboard/blob/main/R/dashboard_modules/05-national_provider_summary.R), which mainly uses shinyGovstyle components, so interested to know if you think that's worth wrapping up in a bigger module in a separate PR, or if there's enough parts already we could just add an example of it in this PR to the showcase app with existing functions? Either way I think it'd be a pattern worth adding to this package to show how it should look (and allow us to easily test it in our upcoming accessibility audit)

rmbielby commented:
> Nice! Honestly quite a neat solution 😄
> 
> To be fair, I've just looked around all over to try to find a reference point for this, and I can't find one anywhere either. So given that, I think I agree with your comment about shinyGovstyle mainly being for styling, so with a few tweaks to the function documentation I'd say this PR could fully resolve #127.
> 
> One other thought I'd had when looking through is that there is another type of download pattern that we could account for - the radio button file type selection with a confirmation button to download, like how EES does it in the table tool? ([example featured table, you'll see if you scroll to the bottom](https://explore-education-statistics.service.gov.uk/data-tables/fast-track/4bf00b2f-dec4-43da-ec12-08dcdd324466)) ![image](https://private-user-images.githubusercontent.com/52536248/400545468-831c09d2-c660-49d9-a090-294c45a833b2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYyNDQyMjcsIm5iZiI6MTczNjI0MzkyNywicGF0aCI6Ii81MjUzNjI0OC80MDA1NDU0NjgtODMxYzA5ZDItYzY2MC00OWQ5LWEwOTAtMjk0YzQ1YTgzM2IyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA3VDA5NTg0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIyYTYxMjMzNGNmZjliNmViZDkzMDRiMDg5NzYyNjdmMWExZTYwMmI4YmQ4YTExYzc2YzU0YTE0NGY3NjZkODImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Z9NrIXZVT1lOWYbcdAe5QKpFKq37C2B6-CLm4oh7PmI)
> 
> I have done a version of this in the [provider dashboard in DfE](https://github.com/dfe-analytical-services/apprenticeships-provider-dashboard/blob/main/R/dashboard_modules/05-national_provider_summary.R), which mainly uses shinyGovstyle components, so interested to know if you think that's worth wrapping up in a bigger module in a separate PR, or if there's enough parts already we could just add an example of it in this PR to the showcase app with existing functions? Either way I think it'd be a pattern worth adding to this package to show how it should look (and allow us to easily test it in our upcoming accessibility audit)

I like the idea of the radio-download module, but instinct is that it should be a separate PR. Happy to do that work (especially as you've already done most of the leg-work in the provider dashboard anyway!), but think it's big / distinct enough it warrants it's own review process.

cjrace reviewed (changes requested):
All looks great, raised #129 and added a comment on #127 for the extra things to look at in other PRs. 

One final thing, as the file size thing still doesn't quite sit right, I think I've worked out what's missing - can we add a warning if the file_size argument is blank? That way if it is blank there's a very clear prompt to add it, and if anyone so desires they can ignore too (though really if they're using this it's probably because they care about best practices so they should expect to be nudged on them)

rmbielby commented:
> All looks great, raised #129 and added a comment on #127 for the extra things to look at in other PRs.
> 
> One final thing, as the file size thing still doesn't quite sit right, I think I've worked out what's missing - can we add a warning if the file_size argument is blank? That way if it is blank there's a very clear prompt to add it, and if anyone so desires they can ignore too (though really if they're using this it's probably because they care about best practices so they should expect to be nudged on them)

I've added a console warning now (and some tests).

cjrace reviewed (approved):
Thanks for adding that!
